From patchwork Wed Aug 28 17:02:55 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jan Dakinevich <jan.dakinevich@virtuozzo.com>
X-Patchwork-Id: 11119487
Return-Path: <SRS0=KojS=WY=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 13B9D14D5
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 28 Aug 2019 17:03:16 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id DF2B72339E
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 28 Aug 2019 17:03:15 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (1024-bit key) header.d=virtuozzo.com header.i=@virtuozzo.com
 header.b="HEJ4FiAx"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726693AbfH1RDB (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 28 Aug 2019 13:03:01 -0400
Received: from mail-eopbgr140095.outbound.protection.outlook.com
 ([40.107.14.95]:48192
        "EHLO EUR01-VE1-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S1726400AbfH1RDA (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 28 Aug 2019 13:03:00 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=O1SgP3pt1F+VplJBID8OtUYTL2SoJYSHOPKvGldbsiKlFAp5b9Xgkf2Mp15Un2JqDUWPBK6VTmOjkLa2Hbv6hGbxQyGfznF2vifKbMPjJRcETozGhYtYdGvvdJTm484ZXzL+3walln30+CQFxKOcmYdDmgfJMq7NScc8Jk38kQ9MYldI+FuyKEtgxSTk4mSZrXJ7ferh151nnhX0SzADcVY9tfjCLwwg+dy45m5iUoesQv20kCy+Ms3OqggeBPhDT/4JvaVdfY6togZLHuwzwtudYwjUU0bw8wkgMcql1EQouklgebf6AgAHaolKIf94e1M6b8sHE9eUuPSAE4Lnvw==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=Wh3jtT+lXogQSO3KQAYL4qcU6PTMONzl/DOKLnMK8IM=;
 b=eq5aUk0VZ1tHDQiwGvzfTdJPoL0lxWcuBjXIqVt+mzPikUwK40oxD4ZLqoEoH26IsIsnO9tc8wL+9Ez0l/rqS4KNVqZpXxc7RxZGoESHCK1DgehNdJApicFucREDAyw8ZjcRaWMKyjlpwX7oHI/vrqpMHOlTBsDb7b3hh4Ff9CnZ6KEY/jIixUTz3laQsJ3QlQWe/gcFgRJFdt8sbVjI3hUlPf/5cvGC3AkA5KfkNFATL4FDi1ACAOTe/PC2WkhshXkoOR6o1cr4jNHnAoQZeTT5kSsLBc+eVS/+G8B5ZW8LymqQHlbYYxYem0nlWTtFU/TQ1biI/uxAZ6jsp7kHQA==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=virtuozzo.com; dmarc=pass action=none
 header.from=virtuozzo.com; dkim=pass header.d=virtuozzo.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=virtuozzo.com;
 s=selector2;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=Wh3jtT+lXogQSO3KQAYL4qcU6PTMONzl/DOKLnMK8IM=;
 b=HEJ4FiAxcObqo2cRAyi/Jkyq2+mZUWfxpHTaKpsFsmsl20oD63L/AU98ROHZ3uyIE4wCbNU3dRso6PvTuuoSn30ryZ+bp1pcIrU8Wbqx759ddpnKgxodEDXKl4QGgpukyiytvRgwkKfQpPHnh85D5OPV2mRKIazQTgaz2WFWcic=
Received: from VI1PR08MB2782.eurprd08.prod.outlook.com (10.170.236.143) by
 VI1PR08MB3181.eurprd08.prod.outlook.com (52.133.15.144) with Microsoft SMTP
 Server (version=TLS1_2, cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id
 15.20.2199.19; Wed, 28 Aug 2019 17:02:55 +0000
Received: from VI1PR08MB2782.eurprd08.prod.outlook.com
 ([fe80::2969:e370:fb70:71a]) by VI1PR08MB2782.eurprd08.prod.outlook.com
 ([fe80::2969:e370:fb70:71a%3]) with mapi id 15.20.2178.023; Wed, 28 Aug 2019
 17:02:55 +0000
From: Jan Dakinevich <jan.dakinevich@virtuozzo.com>
To: "linux-kernel@vger.kernel.org" <linux-kernel@vger.kernel.org>
CC: Jan Dakinevich <jan.dakinevich@virtuozzo.com>,
 Denis Lunev <den@virtuozzo.com>, Roman Kagan <rkagan@virtuozzo.com>,
 Denis Plotnikov <dplotnikov@virtuozzo.com>,
 Paolo Bonzini <pbonzini@redhat.com>,
 =?iso-8859-2?q?Radim_Kr=E8m=E1=F8?= <rkrcmar@redhat.com>,
 Sean Christopherson <sean.j.christopherson@intel.com>,
 Vitaly Kuznetsov <vkuznets@redhat.com>, Wanpeng Li <wanpengli@tencent.com>,
 Jim Mattson <jmattson@google.com>, Joerg Roedel <joro@8bytes.org>,
 Thomas Gleixner <tglx@linutronix.de>, Ingo Molnar <mingo@redhat.com>,
 Borislav Petkov <bp@alien8.de>, "H. Peter Anvin" <hpa@zytor.com>,
 "x86@kernel.org" <x86@kernel.org>,
 "kvm@vger.kernel.org" <kvm@vger.kernel.org>
Subject: [PATCH v2 1/3] KVM: x86: always stop emulation on page fault
Thread-Topic: [PATCH v2 1/3] KVM: x86: always stop emulation on page fault
Thread-Index: AQHVXcJv3UD3akQ2HkiJEaqhuDTLcQ==
Date: Wed, 28 Aug 2019 17:02:55 +0000
Message-ID: <1567011759-9969-2-git-send-email-jan.dakinevich@virtuozzo.com>
References: <1567011759-9969-1-git-send-email-jan.dakinevich@virtuozzo.com>
In-Reply-To: <1567011759-9969-1-git-send-email-jan.dakinevich@virtuozzo.com>
Accept-Language: en-US
Content-Language: en-US
X-MS-Has-Attach: 
X-MS-TNEF-Correlator: 
x-clientproxiedby: HE1PR05CA0154.eurprd05.prod.outlook.com
 (2603:10a6:7:28::41) To VI1PR08MB2782.eurprd08.prod.outlook.com
 (2603:10a6:802:19::15)
authentication-results: spf=none (sender IP is )
 smtp.mailfrom=jan.dakinevich@virtuozzo.com;
x-ms-exchange-messagesentrepresentingtype: 1
x-mailer: git-send-email 2.1.4
x-originating-ip: [185.231.240.5]
x-ms-publictraffictype: Email
x-ms-office365-filtering-correlation-id: b76d02b3-7759-4981-aafe-08d72bd991bd
x-microsoft-antispam: 
 BCL:0;PCL:0;RULEID:(2390118)(7020095)(4652040)(8989299)(4534185)(4627221)(201703031133081)(201702281549075)(8990200)(5600166)(711020)(4605104)(1401327)(2017052603328)(7193020);SRVR:VI1PR08MB3181;
x-ms-traffictypediagnostic: VI1PR08MB3181:
x-ms-exchange-transport-forked: True
x-microsoft-antispam-prvs: 
 <VI1PR08MB3181FC1ABB7F3A3CA1650A218AA30@VI1PR08MB3181.eurprd08.prod.outlook.com>
x-ms-oob-tlc-oobclassifiers: OLM:1284;
x-forefront-prvs: 014304E855
x-forefront-antispam-report: 
 SFV:NSPM;SFS:(10019020)(376002)(366004)(346002)(39850400004)(136003)(396003)(189003)(199004)(478600001)(26005)(4326008)(8936002)(14444005)(186003)(102836004)(3846002)(256004)(476003)(44832011)(6486002)(66066001)(6512007)(305945005)(2906002)(2501003)(446003)(86362001)(6436002)(6116002)(66446008)(8676002)(66946007)(64756008)(66556008)(76176011)(66476007)(316002)(7736002)(81156014)(52116002)(25786009)(81166006)(71190400001)(71200400001)(6506007)(486006)(386003)(14454004)(53936002)(99286004)(2616005)(5660300002)(6916009)(2351001)(54906003)(5640700003)(36756003)(50226002)(11346002)(7416002);DIR:OUT;SFP:1102;SCL:1;SRVR:VI1PR08MB3181;H:VI1PR08MB2782.eurprd08.prod.outlook.com;FPR:;SPF:None;LANG:en;PTR:InfoNoRecords;MX:1;A:1;
received-spf: None (protection.outlook.com: virtuozzo.com does not designate
 permitted sender hosts)
x-ms-exchange-senderadcheck: 1
x-microsoft-antispam-message-info: 
 0MKtYWJkeb3OZ/91KKHEedEMwjJu7z+alGdFdqTmrl/clHEYXaka8grIbTD5SsCEP+fctdOZ2NruFikW3bVCbUXqu/1oavhA9Ccqx1rRqiBQp9HoHiswG0Cu8gRE64GoUFt5Ut9r1jidcRPrc6FlEXOnTvwgytwg9IPSRDEqwna2lFl4lbDm1YhAh7u9z7Z4cj5+6N20XKtux9ob6yQNiurOHbJC896sQx/4AmaeDktsUMW2GOyNklmiSTQvJ+PhEdl6d1XOurtPNAfnabtGtjY5V2Ezp96TKZH1QKtR6Cd+Vmwt5EGGgpmmoM/J+oStyxn+Jik5mDpzNh0xtoCo16wHgrXO9/w64a9YYEWcP/FtmCXYKHM8OCEAdlr8mcJONtiYbJAHVGXGzd1dILBtSxfcYorgHWt+wkWWbcJwyHc=
Content-Type: text/plain; charset="iso-8859-2"
MIME-Version: 1.0
X-OriginatorOrg: virtuozzo.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 b76d02b3-7759-4981-aafe-08d72bd991bd
X-MS-Exchange-CrossTenant-originalarrivaltime: 28 Aug 2019 17:02:55.3276
 (UTC)
X-MS-Exchange-CrossTenant-fromentityheader: Hosted
X-MS-Exchange-CrossTenant-id: 0bc7f26d-0264-416e-a6fc-8352af79c58f
X-MS-Exchange-CrossTenant-mailboxtype: HOSTED
X-MS-Exchange-CrossTenant-userprincipalname: 
 t83KJU1zpEBoq+zSBDdFd//2pN8xrqtCY+Dziha7dgHyXOURf3f1BL9AwlJGbkPS+Rbae747Pe2wp1q0ETBoCt4HjmVhJEovtXuQCT8m+Bw=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: VI1PR08MB3181
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

inject_emulated_exception() returns true if and only if nested page
fault happens. However, page fault can come from guest page tables
walk, either nested or not nested. In both cases we should stop an
attempt to read under RIP and give guest to step over its own page
fault handler.

Fixes: 6ea6e84 ("KVM: x86: inject exceptions produced by x86_decode_insn")
Cc: Denis Lunev <den@virtuozzo.com>
Cc: Roman Kagan <rkagan@virtuozzo.com>
Cc: Denis Plotnikov <dplotnikov@virtuozzo.com>
Signed-off-by: Jan Dakinevich <jan.dakinevich@virtuozzo.com>
---
 arch/x86/kvm/x86.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index b4cfd78..903fb7c 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -6537,8 +6537,10 @@ int x86_emulate_instruction(struct kvm_vcpu *vcpu,
 			if (reexecute_instruction(vcpu, cr2, write_fault_to_spt,
 						emulation_type))
 				return EMULATE_DONE;
-			if (ctxt->have_exception && inject_emulated_exception(vcpu))
+			if (ctxt->have_exception) {
+				inject_emulated_exception(vcpu);
 				return EMULATE_DONE;
+			}
 			if (emulation_type & EMULTYPE_SKIP)
 				return EMULATE_FAIL;
 			return handle_emulation_failure(vcpu, emulation_type);

From patchwork Wed Aug 28 17:02:57 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jan Dakinevich <jan.dakinevich@virtuozzo.com>
X-Patchwork-Id: 11119485
Return-Path: <SRS0=KojS=WY=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 7D63E112C
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 28 Aug 2019 17:03:14 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id 50A6422DA7
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 28 Aug 2019 17:03:14 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (1024-bit key) header.d=virtuozzo.com header.i=@virtuozzo.com
 header.b="Cak57R9m"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726857AbfH1RDE (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 28 Aug 2019 13:03:04 -0400
Received: from mail-eopbgr140095.outbound.protection.outlook.com
 ([40.107.14.95]:48192
        "EHLO EUR01-VE1-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S1726415AbfH1RDC (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 28 Aug 2019 13:03:02 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=KIF6HGIP8GH+jxrahZhXWnA5ybvp4En3ui6t+QWwc5Nh6D6J17Dg0TayB15AQcDHThtkmhi/5kK0ddwe3UDbX4nUmfFED4xM0VqNGYqUJHoR1PGLMCHQbp/uLaQnEBmHYLciihCe37XfcyHKJYEyGvryF9YEhe0oWh/n/CK4Wshb7H/ut7rId332c1N1jfdp8CMjhbkKqIWbOTB1OqS+MsQo4fWupzMNqIXA7puUFHvqoKqxXpjGzX4Kyj2dz2g43f2x0BT3CyakXE1T5/o8MqaLk3GfBI/OMHYvrlQ4pT0DnQIhy2L2x1rb6xz63IUK4c1Tl0JizYf+SaWYT11Jaw==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=QIm7U946QZFxCz9FGcoeYr/NXQH9LC9JmK7WnYfIDCg=;
 b=g4MCycWeYQKLAmedOzwuuyLSS5Cz1N0yZeMZTvsa/cZHedgdytKNE0JSBrXUUmsryuu+VVWqfzRqsZIYKiV8W6zADADqKn/ZgaDuyxp694tm+LgaelnKgHM0CX0PNgxpok5GAeEP042zx/MF0yeAr0N0rkWbmPngwBV+oYVBGIVg7vhjSm0iNARkkcK3LWw3z3+phu9ydtN4n00uDCGh/LsKScuxxbuoZ/qcpIMiSFbzH4vPS6lEQ6/dZsMXIprW1xJVlogMnbRaqvYpfO9PfAc2m+n/ps+QqtWszff2/mj66zDjj7eb7bvrvF0cGsREQ7x/YPZemmo7ZBrEZf4mjQ==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=virtuozzo.com; dmarc=pass action=none
 header.from=virtuozzo.com; dkim=pass header.d=virtuozzo.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=virtuozzo.com;
 s=selector2;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=QIm7U946QZFxCz9FGcoeYr/NXQH9LC9JmK7WnYfIDCg=;
 b=Cak57R9mRqDAPVmdt62QKTmyTqfqQkkacXYQotR0TyznYivL0o6y3ozSVoOxOrLFfl+/dG5dbW/e4dfJjMTr4k95H9tQgbe4VctAlL8jEklnwLvSFMBds4TuhgDD7yA5+t8cm2HOKaj0ehjE0aLgz9rzxuc0wXR9xh1qCezO2s0=
Received: from VI1PR08MB2782.eurprd08.prod.outlook.com (10.170.236.143) by
 VI1PR08MB3181.eurprd08.prod.outlook.com (52.133.15.144) with Microsoft SMTP
 Server (version=TLS1_2, cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id
 15.20.2199.19; Wed, 28 Aug 2019 17:02:57 +0000
Received: from VI1PR08MB2782.eurprd08.prod.outlook.com
 ([fe80::2969:e370:fb70:71a]) by VI1PR08MB2782.eurprd08.prod.outlook.com
 ([fe80::2969:e370:fb70:71a%3]) with mapi id 15.20.2178.023; Wed, 28 Aug 2019
 17:02:57 +0000
From: Jan Dakinevich <jan.dakinevich@virtuozzo.com>
To: "linux-kernel@vger.kernel.org" <linux-kernel@vger.kernel.org>
CC: Jan Dakinevich <jan.dakinevich@virtuozzo.com>,
 Denis Lunev <den@virtuozzo.com>, Roman Kagan <rkagan@virtuozzo.com>,
 Denis Plotnikov <dplotnikov@virtuozzo.com>,
 Paolo Bonzini <pbonzini@redhat.com>,
 =?iso-8859-2?q?Radim_Kr=E8m=E1=F8?= <rkrcmar@redhat.com>,
 Sean Christopherson <sean.j.christopherson@intel.com>,
 Vitaly Kuznetsov <vkuznets@redhat.com>, Wanpeng Li <wanpengli@tencent.com>,
 Jim Mattson <jmattson@google.com>, Joerg Roedel <joro@8bytes.org>,
 Thomas Gleixner <tglx@linutronix.de>, Ingo Molnar <mingo@redhat.com>,
 Borislav Petkov <bp@alien8.de>, "H. Peter Anvin" <hpa@zytor.com>,
 "x86@kernel.org" <x86@kernel.org>,
 "kvm@vger.kernel.org" <kvm@vger.kernel.org>
Subject: [PATCH v2 2/3] KVM: x86: make exception_class() and exception_type()
 globally visible
Thread-Topic: [PATCH v2 2/3] KVM: x86: make exception_class() and
 exception_type() globally visible
Thread-Index: AQHVXcJwZ8H/6OmPEECkMYgQWBHwtg==
Date: Wed, 28 Aug 2019 17:02:57 +0000
Message-ID: <1567011759-9969-3-git-send-email-jan.dakinevich@virtuozzo.com>
References: <1567011759-9969-1-git-send-email-jan.dakinevich@virtuozzo.com>
In-Reply-To: <1567011759-9969-1-git-send-email-jan.dakinevich@virtuozzo.com>
Accept-Language: en-US
Content-Language: en-US
X-MS-Has-Attach: 
X-MS-TNEF-Correlator: 
x-clientproxiedby: HE1PR05CA0154.eurprd05.prod.outlook.com
 (2603:10a6:7:28::41) To VI1PR08MB2782.eurprd08.prod.outlook.com
 (2603:10a6:802:19::15)
authentication-results: spf=none (sender IP is )
 smtp.mailfrom=jan.dakinevich@virtuozzo.com;
x-ms-exchange-messagesentrepresentingtype: 1
x-mailer: git-send-email 2.1.4
x-originating-ip: [185.231.240.5]
x-ms-publictraffictype: Email
x-ms-office365-filtering-correlation-id: 1f0e28ce-82ee-4ea6-d7b4-08d72bd99317
x-microsoft-antispam: 
 BCL:0;PCL:0;RULEID:(2390118)(7020095)(4652040)(8989299)(4534185)(4627221)(201703031133081)(201702281549075)(8990200)(5600166)(711020)(4605104)(1401327)(2017052603328)(7193020);SRVR:VI1PR08MB3181;
x-ms-traffictypediagnostic: VI1PR08MB3181:
x-ms-exchange-transport-forked: True
x-microsoft-antispam-prvs: 
 <VI1PR08MB3181217EF23CFF0E7A3C5C9D8AA30@VI1PR08MB3181.eurprd08.prod.outlook.com>
x-ms-oob-tlc-oobclassifiers: OLM:4125;
x-forefront-prvs: 014304E855
x-forefront-antispam-report: 
 SFV:NSPM;SFS:(10019020)(376002)(366004)(346002)(39850400004)(136003)(396003)(189003)(199004)(478600001)(26005)(4326008)(8936002)(186003)(102836004)(3846002)(256004)(476003)(44832011)(6486002)(66066001)(6512007)(305945005)(2906002)(2501003)(446003)(86362001)(6436002)(6116002)(66446008)(8676002)(66946007)(64756008)(66556008)(76176011)(66476007)(316002)(7736002)(81156014)(52116002)(25786009)(81166006)(71190400001)(71200400001)(6506007)(486006)(386003)(14454004)(53936002)(99286004)(2616005)(5660300002)(6916009)(2351001)(54906003)(5640700003)(36756003)(50226002)(11346002)(7416002);DIR:OUT;SFP:1102;SCL:1;SRVR:VI1PR08MB3181;H:VI1PR08MB2782.eurprd08.prod.outlook.com;FPR:;SPF:None;LANG:en;PTR:InfoNoRecords;MX:1;A:1;
received-spf: None (protection.outlook.com: virtuozzo.com does not designate
 permitted sender hosts)
x-ms-exchange-senderadcheck: 1
x-microsoft-antispam-message-info: 
 4HqlzTY/k8eD1XZTzzeUCt10LcolWzUiCbl5NDQd6sUON84aOW4lYiaMe6PQQmGnwlWOg/WNaRz0/6Lp6D8yXGuw80jauPwrYpLoWwFUoXO6WO6QNAGUPdbw9eX7StGLHmyEcySNFV84enxCPvhg6lyXL0YSrhPx2y6Gxdceu4x8RTvpI/pjj/FIQPx1HAvXReBI3cKkubXEwQzAJa51sRgRQ9zbNi3k9GKcTJI7/InGtq0fwUi1/mTFqnjOCI0XcQzheveZl58JpnuEdwGaZbvjn/UECKeXV57q5M5zGQxc0R0AknoUQaXqezp5WWdd3W4MXGz8RrWM6cB0dh8VfRXuRWt3Rz3pOiZquOeO0tWdlWy9Q1fV9hp/IDhUPPPrg23kyB8Q2mL1kdYXP/zlz+Pn0zHXJkeGID+0BJfM3U0=
Content-Type: text/plain; charset="iso-8859-2"
MIME-Version: 1.0
X-OriginatorOrg: virtuozzo.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 1f0e28ce-82ee-4ea6-d7b4-08d72bd99317
X-MS-Exchange-CrossTenant-originalarrivaltime: 28 Aug 2019 17:02:57.4423
 (UTC)
X-MS-Exchange-CrossTenant-fromentityheader: Hosted
X-MS-Exchange-CrossTenant-id: 0bc7f26d-0264-416e-a6fc-8352af79c58f
X-MS-Exchange-CrossTenant-mailboxtype: HOSTED
X-MS-Exchange-CrossTenant-userprincipalname: 
 38SDz5KyO+Jv3UMRXJJ43KtbOu5vHVkL1WQNl98u30Fb2ilqbqNp7zamB85bt58Dnf4z12RVaBFD++/kKFNn8gV2T0LLbs2o221F/0psFDU=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: VI1PR08MB3181
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

exception_type() function was moved for upcoming sanity check in
emulation code. exceptions_class() function is not supposed to be used
right now, but it was moved as well to keep things together.

Cc: Denis Lunev <den@virtuozzo.com>
Cc: Roman Kagan <rkagan@virtuozzo.com>
Cc: Denis Plotnikov <dplotnikov@virtuozzo.com>
Signed-off-by: Jan Dakinevich <jan.dakinevich@virtuozzo.com>
---
 arch/x86/kvm/x86.c | 46 ----------------------------------------------
 arch/x86/kvm/x86.h | 46 ++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 46 insertions(+), 46 deletions(-)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 903fb7c..2b69ae0 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -364,52 +364,6 @@ asmlinkage __visible void kvm_spurious_fault(void)
 }
 EXPORT_SYMBOL_GPL(kvm_spurious_fault);
 
-#define EXCPT_BENIGN		0
-#define EXCPT_CONTRIBUTORY	1
-#define EXCPT_PF		2
-
-static int exception_class(int vector)
-{
-	switch (vector) {
-	case PF_VECTOR:
-		return EXCPT_PF;
-	case DE_VECTOR:
-	case TS_VECTOR:
-	case NP_VECTOR:
-	case SS_VECTOR:
-	case GP_VECTOR:
-		return EXCPT_CONTRIBUTORY;
-	default:
-		break;
-	}
-	return EXCPT_BENIGN;
-}
-
-#define EXCPT_FAULT		0
-#define EXCPT_TRAP		1
-#define EXCPT_ABORT		2
-#define EXCPT_INTERRUPT		3
-
-static int exception_type(int vector)
-{
-	unsigned int mask;
-
-	if (WARN_ON(vector > 31 || vector == NMI_VECTOR))
-		return EXCPT_INTERRUPT;
-
-	mask = 1 << vector;
-
-	/* #DB is trap, as instruction watchpoints are handled elsewhere */
-	if (mask & ((1 << DB_VECTOR) | (1 << BP_VECTOR) | (1 << OF_VECTOR)))
-		return EXCPT_TRAP;
-
-	if (mask & ((1 << DF_VECTOR) | (1 << MC_VECTOR)))
-		return EXCPT_ABORT;
-
-	/* Reserved exceptions will result in fault */
-	return EXCPT_FAULT;
-}
-
 void kvm_deliver_exception_payload(struct kvm_vcpu *vcpu)
 {
 	unsigned nr = vcpu->arch.exception.nr;
diff --git a/arch/x86/kvm/x86.h b/arch/x86/kvm/x86.h
index b5274e2..2b66347 100644
--- a/arch/x86/kvm/x86.h
+++ b/arch/x86/kvm/x86.h
@@ -369,4 +369,50 @@ static inline bool kvm_pat_valid(u64 data)
 void kvm_load_guest_xcr0(struct kvm_vcpu *vcpu);
 void kvm_put_guest_xcr0(struct kvm_vcpu *vcpu);
 
+#define EXCPT_BENIGN		0
+#define EXCPT_CONTRIBUTORY	1
+#define EXCPT_PF		2
+
+static inline int exception_class(int vector)
+{
+	switch (vector) {
+	case PF_VECTOR:
+		return EXCPT_PF;
+	case DE_VECTOR:
+	case TS_VECTOR:
+	case NP_VECTOR:
+	case SS_VECTOR:
+	case GP_VECTOR:
+		return EXCPT_CONTRIBUTORY;
+	default:
+		break;
+	}
+	return EXCPT_BENIGN;
+}
+
+#define EXCPT_FAULT		0
+#define EXCPT_TRAP		1
+#define EXCPT_ABORT		2
+#define EXCPT_INTERRUPT		3
+
+static inline int exception_type(int vector)
+{
+	unsigned int mask;
+
+	if (WARN_ON(vector > 31 || vector == NMI_VECTOR))
+		return EXCPT_INTERRUPT;
+
+	mask = 1 << vector;
+
+	/* #DB is trap, as instruction watchpoints are handled elsewhere */
+	if (mask & ((1 << DB_VECTOR) | (1 << BP_VECTOR) | (1 << OF_VECTOR)))
+		return EXCPT_TRAP;
+
+	if (mask & ((1 << DF_VECTOR) | (1 << MC_VECTOR)))
+		return EXCPT_ABORT;
+
+	/* Reserved exceptions will result in fault */
+	return EXCPT_FAULT;
+}
+
 #endif

From patchwork Wed Aug 28 17:02:59 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jan Dakinevich <jan.dakinevich@virtuozzo.com>
X-Patchwork-Id: 11119483
Return-Path: <SRS0=KojS=WY=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 420B9112C
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 28 Aug 2019 17:03:09 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id 16335233FF
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed, 28 Aug 2019 17:03:09 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (1024-bit key) header.d=virtuozzo.com header.i=@virtuozzo.com
 header.b="DFG79O+v"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726882AbfH1RDI (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 28 Aug 2019 13:03:08 -0400
Received: from mail-eopbgr140095.outbound.protection.outlook.com
 ([40.107.14.95]:48192
        "EHLO EUR01-VE1-obe.outbound.protection.outlook.com"
        rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
        id S1726837AbfH1RDF (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 28 Aug 2019 13:03:05 -0400
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=jUJLWoKMmzmjMJ3MYBHkfxkc6oh6SxW4mdmVl7/jWfepDBbb1mEGck+7W1F8nILZCBlio5a/5NjkLqOFyDN2ewASw23W4ybujOkuGwPg7FJumV0iWQYWZyHluMFNYlpr1c3m2QqveYGNt9TzpSpminnLZOd31RQgrPzQcYCJbdZcQ/FxH3u+8N27acrUW25TuXvq1HxKMaHOnFd40SHpv+PQ0XgdV0yBLwfCoQK46FcegviS4k4EW8iSNLa9a7yAjtrN1wQ6FSOhWxkQ722eX1XkSoba3z6tNn7gEVeiTQ1+xTPpA8wfEHN4meXlqdBIUt1fCVDRyRQZUTbSH4lcXw==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=FVE+tKSCP26l93DhJV3K+1nKtyODUHjKA+FnaFcYU5U=;
 b=nVq8w/X5MKOPp65xiQZnNEdvtiTkZ89KGnoP37YAgEZbZHxHLbxfp5ytUK4G/GMtadl4eqoz7Ww2tzbfEKXKtL+Asg6z0EhFZUWAA7PpDy01PVAbwr6MmDuWg2bN2lJwOhUAx5XKhpv43surkZN0wFrM46PYNz/GMCQM/ydlabYDczulHByoEfPrJjpCwJ+cad9H/q+Buq6UB2w0cuCPB6aMgTwPtxp4KZ0tPxq2AxGcrnVJJkAJnTP8ggHHrkcfWCMXgCz9yaZt4fEAPcuo1GjDqYH/kJfSKjYposySjMKII/65gric3npgoK95/AYRo5Cy+6tfyikH4ykJ0Moytg==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=virtuozzo.com; dmarc=pass action=none
 header.from=virtuozzo.com; dkim=pass header.d=virtuozzo.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=virtuozzo.com;
 s=selector2;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=FVE+tKSCP26l93DhJV3K+1nKtyODUHjKA+FnaFcYU5U=;
 b=DFG79O+vtUwhq+SW1k1JU62KGVHrW7zcVXD46MsVFUTCTG6AN/vTkAj/rvupGzgHGv7BO/hJgOS3d/8QnhIwFt9922/RHtYKKbYEVvVJf3L/ifHB5mhzUw4ia67PC62R+n0IFr0JUm/WekJjkNp9CiY84hixmjdIcZ9uEtVbkYU=
Received: from VI1PR08MB2782.eurprd08.prod.outlook.com (10.170.236.143) by
 VI1PR08MB3181.eurprd08.prod.outlook.com (52.133.15.144) with Microsoft SMTP
 Server (version=TLS1_2, cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id
 15.20.2199.19; Wed, 28 Aug 2019 17:02:59 +0000
Received: from VI1PR08MB2782.eurprd08.prod.outlook.com
 ([fe80::2969:e370:fb70:71a]) by VI1PR08MB2782.eurprd08.prod.outlook.com
 ([fe80::2969:e370:fb70:71a%3]) with mapi id 15.20.2178.023; Wed, 28 Aug 2019
 17:02:59 +0000
From: Jan Dakinevich <jan.dakinevich@virtuozzo.com>
To: "linux-kernel@vger.kernel.org" <linux-kernel@vger.kernel.org>
CC: Jan Dakinevich <jan.dakinevich@virtuozzo.com>,
 Denis Lunev <den@virtuozzo.com>, Roman Kagan <rkagan@virtuozzo.com>,
 Denis Plotnikov <dplotnikov@virtuozzo.com>,
 Paolo Bonzini <pbonzini@redhat.com>,
 =?iso-8859-2?q?Radim_Kr=E8m=E1=F8?= <rkrcmar@redhat.com>,
 Sean Christopherson <sean.j.christopherson@intel.com>,
 Vitaly Kuznetsov <vkuznets@redhat.com>, Wanpeng Li <wanpengli@tencent.com>,
 Jim Mattson <jmattson@google.com>, Joerg Roedel <joro@8bytes.org>,
 Thomas Gleixner <tglx@linutronix.de>, Ingo Molnar <mingo@redhat.com>,
 Borislav Petkov <bp@alien8.de>, "H. Peter Anvin" <hpa@zytor.com>,
 "x86@kernel.org" <x86@kernel.org>,
 "kvm@vger.kernel.org" <kvm@vger.kernel.org>
Subject: [PATCH v2 3/3] KVM: x86: set ctxt->have_exception in
 x86_decode_insn()
Thread-Topic: [PATCH v2 3/3] KVM: x86: set ctxt->have_exception in
 x86_decode_insn()
Thread-Index: AQHVXcJxnnwQLIGeYUSYlBg22R5cuw==
Date: Wed, 28 Aug 2019 17:02:59 +0000
Message-ID: <1567011759-9969-4-git-send-email-jan.dakinevich@virtuozzo.com>
References: <1567011759-9969-1-git-send-email-jan.dakinevich@virtuozzo.com>
In-Reply-To: <1567011759-9969-1-git-send-email-jan.dakinevich@virtuozzo.com>
Accept-Language: en-US
Content-Language: en-US
X-MS-Has-Attach: 
X-MS-TNEF-Correlator: 
x-clientproxiedby: HE1PR05CA0154.eurprd05.prod.outlook.com
 (2603:10a6:7:28::41) To VI1PR08MB2782.eurprd08.prod.outlook.com
 (2603:10a6:802:19::15)
authentication-results: spf=none (sender IP is )
 smtp.mailfrom=jan.dakinevich@virtuozzo.com;
x-ms-exchange-messagesentrepresentingtype: 1
x-mailer: git-send-email 2.1.4
x-originating-ip: [185.231.240.5]
x-ms-publictraffictype: Email
x-ms-office365-filtering-correlation-id: 96a8620c-ffc4-4c78-e996-08d72bd99440
x-microsoft-antispam: 
 BCL:0;PCL:0;RULEID:(2390118)(7020095)(4652040)(8989299)(4534185)(4627221)(201703031133081)(201702281549075)(8990200)(5600166)(711020)(4605104)(1401327)(2017052603328)(7193020);SRVR:VI1PR08MB3181;
x-ms-traffictypediagnostic: VI1PR08MB3181:
x-ms-exchange-transport-forked: True
x-microsoft-antispam-prvs: 
 <VI1PR08MB3181536D61CFF41CEC2DD7998AA30@VI1PR08MB3181.eurprd08.prod.outlook.com>
x-ms-oob-tlc-oobclassifiers: OLM:2733;
x-forefront-prvs: 014304E855
x-forefront-antispam-report: 
 SFV:NSPM;SFS:(10019020)(376002)(366004)(346002)(39850400004)(136003)(396003)(189003)(199004)(478600001)(26005)(4326008)(8936002)(14444005)(186003)(102836004)(3846002)(256004)(476003)(44832011)(6486002)(66066001)(6512007)(305945005)(4744005)(2906002)(2501003)(446003)(86362001)(6436002)(6116002)(66446008)(8676002)(66946007)(64756008)(66556008)(76176011)(66476007)(316002)(7736002)(81156014)(52116002)(25786009)(81166006)(71190400001)(71200400001)(6506007)(486006)(386003)(14454004)(53936002)(99286004)(2616005)(5660300002)(6916009)(2351001)(54906003)(5640700003)(36756003)(50226002)(11346002)(7416002);DIR:OUT;SFP:1102;SCL:1;SRVR:VI1PR08MB3181;H:VI1PR08MB2782.eurprd08.prod.outlook.com;FPR:;SPF:None;LANG:en;PTR:InfoNoRecords;MX:1;A:1;
received-spf: None (protection.outlook.com: virtuozzo.com does not designate
 permitted sender hosts)
x-ms-exchange-senderadcheck: 1
x-microsoft-antispam-message-info: 
 2gaLlwxtNvt9Zytn3/lKCmB93uhWT5fkGLpL/GCZDsvhLcNBwrWL0Ym5i5iGfqTuFEOnQDcLUCnvsRlilg8tDECIdHaPlAYdn6xeoVTo7L4fyTeRRkrZRT8MuDM4M0oU/OT/q54ANg99tN+wtcH5shkHatjD2eu5Jo1KPiNzUhkjAOSXu4pMey5Ak7JA9R9UgWsqaecjiJMuX5VMNvi/pSA3FdEp70biURZIv1L9G3r24YdUoxMCKoBv/Lt+hzD1SlrLS9ZH5jHhvS6rFxfYmgPFykQW1F8S5d37eFkvdMYyw/FmYYTtfqBmQBXPkvRSXAIigd00+ZjW2KtFPb4YWuUoqdi9JxhfGCSxDcKL03UrLIqFz+V46SXye1wDpUcjk/n0W6xC/iz5UNuxj1ULrxWMsV2ODQ5mTySAv60U9lk=
Content-Type: text/plain; charset="iso-8859-2"
MIME-Version: 1.0
X-OriginatorOrg: virtuozzo.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 96a8620c-ffc4-4c78-e996-08d72bd99440
X-MS-Exchange-CrossTenant-originalarrivaltime: 28 Aug 2019 17:02:59.5321
 (UTC)
X-MS-Exchange-CrossTenant-fromentityheader: Hosted
X-MS-Exchange-CrossTenant-id: 0bc7f26d-0264-416e-a6fc-8352af79c58f
X-MS-Exchange-CrossTenant-mailboxtype: HOSTED
X-MS-Exchange-CrossTenant-userprincipalname: 
 DPGYn7KqLQz9MV02v8mNpCsTr5LqvI0JLjeycMfdFUq3XfLEMWgvzjFoNz+bLTinRkZBSVE94N3WNzpwBoCLpoLBO+jdycvGm7JZCTl3c7M=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: VI1PR08MB3181
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

x86_emulate_instruction() takes into account ctxt->have_exception flag
during instruction decoding, but in practice this flag is never set in
x86_decode_insn().

Fixes: 6ea6e84 ("KVM: x86: inject exceptions produced by x86_decode_insn")
Cc: Denis Lunev <den@virtuozzo.com>
Cc: Roman Kagan <rkagan@virtuozzo.com>
Cc: Denis Plotnikov <dplotnikov@virtuozzo.com>
Signed-off-by: Jan Dakinevich <jan.dakinevich@virtuozzo.com>
---
 arch/x86/kvm/emulate.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/arch/x86/kvm/emulate.c b/arch/x86/kvm/emulate.c
index bef3c3c..74b4d79 100644
--- a/arch/x86/kvm/emulate.c
+++ b/arch/x86/kvm/emulate.c
@@ -5416,6 +5416,11 @@ int x86_decode_insn(struct x86_emulate_ctxt *ctxt, void *insn, int insn_len)
 					ctxt->memopp->addr.mem.ea + ctxt->_eip);
 
 done:
+	if (rc == X86EMUL_PROPAGATE_FAULT) {
+		WARN_ON_ONCE(ctxt->exception.vector == UD_VECTOR ||
+			     exception_type(ctxt->exception.vector) == EXCPT_TRAP);
+		ctxt->have_exception = true;
+	}
 	return (rc != X86EMUL_CONTINUE) ? EMULATION_FAILED : EMULATION_OK;
 }
 
